File: m03_robot_description/m03_tf2_with_py/m03_tf2_with_py/turtle_broadcaster.py
# !/usr/bin/env python3

# --------- Import definitions present in other codes of the package ----------
import math
import sys
import numpy as np

# ------------------------ ROS2 rclpy implementations -------------------------
import rclpy
from rclpy.node import Node

# ----------------------- ROS2 Messages and tf2 related -----------------------
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster
from turtlesim.msg import Pose

def quaternion_from_euler(ai, aj, ak):
    """
    Mathematic function that calculates the equivalent quaternion by
    considering raw, pitch and yaw.
    
    Params
    ---
    ai : float
        Raw angle or the rotation considering unit i vector.
    aj : float
        Pitch angle or the rotation considering unit j vector.
    ak : float
        Yaw angle or the rotation considering unit k vector.
    
    Returns
    ---
    q : float array
        Quaternion as an array of four components (x, y, z, w).
    """
    # Component extraction and consideration
    ai /= 2.0
    aj /= 2.0
    ak /= 2.0
    ci = math.cos(ai)
    si = math.sin(ai)
    cj = math.cos(aj)
    sj = math.sin(aj)
    ck = math.cos(ak)
    sk = math.sin(ak)
    cc = ci*ck
    cs = ci*sk
    sc = si*ck
    ss = si*sk
    
    # Quaternion definition
    q = np.empty((4, ))
    q[0] = cj*sc - sj*cs
    q[1] = cj*ss + sj*cc
    q[2] = cj*cs - sj*sc
    q[3] = cj*cc + sj*ss
    return q

# --------------- Base class of the program for tf2 broadcaster ---------------
class TurtlePoseBroad(Node):
    """
    Class oriented to boradcasting the pose (as a tf) of a turtle in turtlesim.
    
    Attributes
    ---
    turtlename : ros2 parameter
        String param that refers to the name of the turtle to follow.
        
    tf_broadcaster : Transform broadcaster
        The broadcaster that will be used to transmit the tf of the turtle.
        
    subscription : subscriber
        Subscribe to turtlesim pose of the turtle with the given turtlename
        
    Methods
    ---
    handle_turtle_pose(msg)
        Callback of the subscription that will boradcast the pose of the
        given turtle (tf).
    """
    def __init__(self):
        """ 
        Constructor that initialize a node with the name 
        'turtle_tf2_frame_publisher', declares a parameter related with turtle
        names (future relation with turtlesim), instance a broadcaster and 
        subscribes to the turtle pose.
        """
        # Initialize node
        super().__init__('turtle_tf2_frame_publisher')

        # Declare and acquire `turtlename` parameter for using with turtlesim
        self.turtlename = self.declare_parameter(
          'turtlename', 'turtle').get_parameter_value().string_value

        # Instance a broadcaster
        self.tf_broadcaster = TransformBroadcaster(self)

        # Subscribe to the turtle pose
        self.subscription = self.create_subscription(
            Pose,                           # Msg type
            f'/{self.turtlename}/pose',     # Topic
            self.handle_turtle_pose,        # Callback
            1)                              # Queue

    def handle_turtle_pose(self, msg):
        # Instance transform
        t = TransformStamped()

        # Time stamp and tf relation 
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'world'
        t.child_frame_id = self.turtlename

        # Getting turtle linear and angular position (2D consideration)
        t.transform.translation.x = msg.x
        t.transform.translation.y = msg.y
        t.transform.translation.z = 0.0

        # As the turtle only rotates relative to z, we obtain the quaternion
        q = quaternion_from_euler(0, 0, msg.theta)
        t.transform.rotation.x = q[0]
        t.transform.rotation.y = q[1]
        t.transform.rotation.z = q[2]
        t.transform.rotation.w = q[3]

        # Send the transformation
        self.tf_broadcaster.sendTransform(t)


def main():
    # Initialize ROS Client Library for Python
    rclpy.init()
    
    # Create a logger for displaying info
    logger = rclpy.logging.get_logger('logger')
    logger.info("Initialized: Broadcaster of turtles tfs.")
    
    # Instance a node 
    node = TurtlePoseBroad()
    
    # Spin the node until Keyboard Interrupt is received
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass

    # Shutdown and close
    logger.info("Closing: Broadcaster of turtles tfs")
    rclpy.shutdown()

File: m03_robot_description/m03_tf2_with_cpp/src/turtle_broadcaster.cpp
// ---------------------- REQUIRED STANDARD HEADERS ---------------------------
#include <functional>  // For function object and standard hash function
#include <memory>      // Dynamic memory management
#include <sstream>     // String stream
#include <string>      // Collection of capabilites for string usage

// ----------------------- ROS2 REQUIRED HEADERS -----------------------------
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_ros/transform_broadcaster.h"

// ----------------------- ROS2 MSGS DEPENDENCIES ----------------------------
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "turtlesim/msg/pose.hpp"

// ----------------------- NAMESPACE CONSIDERATION --------------------------- 
using namespace std::placeholders; // Argument placement

// ------------------ BROADCASTER CLASS IMPLEMENTATION ------------------------
/**
 * Classs that broadcast the tf of a turtle (dynamic) in the turtlesim world,
 * so it allows to obtain info of its position.
 */
class TurtlePoseBroad : public rclcpp::Node
{
public:
    /**
     * User defined constructor that initalize a node with the name 
     * "turtle_tf2_frame_publisher". Also, it declares a turtle param, a tf
     * broadcaster and subscription to a poser.
     */
    TurtlePoseBroad()
        : Node("turtle_tf2_frame_publisher")
    {
        // Parameter declaration
        turtlename_ = this->declare_parameter<std::string>(
            "turtlename", "turtle");
        
        // Definition of tf broadcaster by using unique_ptr
        tf_broad_ = 
            std::make_unique<tf2_ros::TransformBroadcaster>(*this);

        // Obtain the corresponding topic pose by considering turtle name param
        std::ostringstream stream;
        stream << "/" << turtlename_.c_str() << "/pose";
        std::string topic_name = stream.str();

        // Create subscription and link handler of the pose to broadcast tf
        subs_ = this->create_subscription<turtlesim::msg::Pose>(
            topic_name, 10, 
            std::bind(&TurtlePoseBroad::handle_turtle_pose, this, _1));
    }

private:
    /**
     * Function that receives the msg pose of the turtle and broadcast it with
     * tfs.
     * 
     * @param msg Pose msg received of the given turtle
     */
    void handle_turtle_pose(const std::shared_ptr<turtlesim::msg::Pose> msg)
    {
        // Declare transfor stamped message
        geometry_msgs::msg::TransformStamped t_stamp;

        // Add stamp (time, child frame and parent frame)
        t_stamp.header.stamp = this->get_clock()->now();
        t_stamp.header.frame_id = "world";
        t_stamp.child_frame_id = turtlename_.c_str();

        // Add traslational info of the turtle (considering just 2D components)
        t_stamp.transform.translation.x = msg->x;
        t_stamp.transform.translation.y = msg->y;
        t_stamp.transform.translation.z = 0.0;

        // Traslate RPY to quaternion to obtain the given position
        // Herez z is just considered as a 2D rotation axis
        tf2::Quaternion q;
        q.setRPY(0, 0, msg->theta);

        // Add the rotational info of the turtle
        t_stamp.transform.rotation.x = q.x();
        t_stamp.transform.rotation.y = q.y();
        t_stamp.transform.rotation.z = q.z();
        t_stamp.transform.rotation.w = q.w();

        // Broadcast transform
        tf_broad_->sendTransform(t_stamp);
    }

    // Declare subscription via shared_ptr
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subs_;

    // Declare tf broadcaster via unique_ptr
    std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broad_;

    // Add string for turtlename
    std::string turtlename_;
};

// ------------------------- MAIN IMPLMENTATION -------------------------------
int main(int argc, char * argv[])
{
    // Initialize ROS2 Client Library for C++
    rclcpp::init(argc, argv);

    // Spin node by using shared pointer
    rclcpp::spin(std::make_shared<TurtlePoseBroad>());
    
    // Close and shutdown when the spin ends
    rclcpp::shutdown();
    return 0;
}